source/elements/oneMKL/source/domains/spblas/data_types/set_coo_matrix_data.rst

```diff
@@ -0,0 +1,225 @@
+
+   }
+
+.. container:: section
+
+   .. rubric:: Template parameters
+
+   dataType
+      See :ref:`supported template types<onemkl_sparse_supported_types>`. Can be
+      a different type than what was used when creating the ``matrix_handle_t``.
+
+   indexType
+      See :ref:`supported template types<onemkl_sparse_supported_types>`. Can be
+      a different type than what was used when creating the ``matrix_handle_t``.
+
+.. container:: section
+
+   .. rubric:: Input parameters
+
+   queue
+      The SYCL command queue which will be used for SYCL kernels execution.
+
+   smhandle
+      Handle already initialized with :ref:`onemkl_sparse_init_coo_matrix`.
+
+   num_rows
+      Number of rows of the provided data ``val``. Must be at least 0.
+
+   num_cols
+      Number of columns of the provided data ``val``. Must be at least 0.
+
+   nnz
+      The number of explicit entries, also known as Number of Non-Zero elements.
+      Must be at least 0.
+
+   index
+      Indicates how input arrays are indexed. The possible options are described
+      in :ref:`onemkl_enum_index_base` enum class.
+
+   row_ind
+      Buffer of length at least ``nnz`` containing the row indices in
+      ``index``-based numbering. Refer to :ref:`onemkl_sparse_coo` format for
+      detailed description of ``row_ind``.
+
+   col_ind
+      Buffer of length at least ``nnz`` containing the column indices in
+      ``index``-based numbering. Refer to :ref:`onemkl_sparse_coo` format for
+      detailed description of ``col_ind``.
+
+   val
+      Buffer of length at least ``nnz``. Contains the data of the input matrix
+      which is not implicitly zero. The remaining input values are implicit
+      zeros. Refer to :ref:`onemkl_sparse_coo` format for detailed description
+      of ``val``.
+
+.. container:: section
+
+   .. rubric:: Notes
+
+   - The parameters ``num_rows``, ``num_cols`` and ``nnz`` may be zero if and
+     only if ``row_ind``, ``col_ind`` and ``val`` are zero-sized, otherwise they
+     must be strictly greater than zero.
+
+.. container:: section
+
+   .. rubric:: Throws
+
+   This function shall throw the following exceptions if the associated
+   condition is detected. An implementation may throw additional
+   implementation-specific exception(s) in case of error conditions not covered
+   here.
+
+   | :ref:`oneapi::mkl::invalid_argument<onemkl_exception_invalid_argument>`
+   | :ref:`oneapi::mkl::unimplemented<onemkl_exception_unimplemented>`
+   | :ref:`oneapi::mkl::uninitialized<onemkl_exception_uninitialized>`
+   | :ref:`oneapi::mkl::unsupported_device<onemkl_exception_unsupported_device>`
+
+.. _onemkl_sparse_set_coo_matrix_data_usm:
+
+set_coo_matrix_data (USM version)
+---------------------------------
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       template <typename dataType, typename indexType>
+       void set_coo_matrix_data (sycl::queue                          &queue,
```

Comment:
what is the difference between init_coo_matrix and set_coo_matrix_data ?  And should this set case then support asynchronous DAG modification ?  or is it also blocking ?

Reply:
since queue is there, I presume it may have kernels called inside it which would prevent it from being embedded in a host_task to get asynchronous update ...

Reply:
same question for CSR set/init

Reply:
The `init_*` function initializes the structure with a host allocation and sets the data while `set_*` only sets the data. They are both blocking right now.

I don't see in which case we would need `set_*` functions to be asynchronous since they do not use the data, they only update the handles on the host side. I think having asynchronous `init_*` or `set_*` functions is not practical since the user would not know when the handles are updated so they would need to wait before using a handle that has been initialized or reset.

The queue does not necessarily mean that it is needed to submit kernels or host_tasks. Every functions in oneMKL need the queue to select at runtime which backend to dispatch to. Currently we also need it to support the MKLCPU and MKLGPU backends with 2024.1.

Reply:
I am thinking about the potential future case of SpGEMM  C = A * B  sparse * sparse matrix multiplication,  but omatadd is also a place where this could show up

1. init A and B with data
2. create empty C sparse handle
3. call spgemm_compute( q, A,B,C)  to do some work to  determine size of C
4. query size of C   `index_type nnz = sgemm_get_nnz(q, A,B,C)`    I suppose that this might need to involve futures to get nnz value asynchronously ( but if if we never do that then it won't matter if set is asynchronous or not)
5. allocate C arrays and fill it  (could be done with asynchronous allocations to keep chain going)
6. call spgemm_finalize(q, A,B,C) to fill C matrix

it all depends on whether we eventually will have a way to get nnz value and allocate asynchronously or not, to affect whether we need an asynchronous set_xyz_matrix_data() function ...  maybe we will never do that, so this is fine.

Reply:
Right so this would be part of this separate API as well.

As you said being able to allocate asynchronously is not support in SYCL yet as far as I know. Being able to allocate asynchronously with the size being a future type is probably even harder!

Reply:
ok, if the time comes that we get such features, we will be able to address this .  for now, it is fine as is